app-experimental.tex

```diff
@@ -2395,3 +2395,222 @@ \section{Distributed Capabilities For Peripherals And Accelerators} % <<<
+
+Another improvement might be to add a bit that determines whether a capability is allowed to be used within CID spaces.
+This would add fine-grained control over which capabilities can be shared within CID spaces and which are private to the owning compartment.
+
+\subsection{Code and Data Compartments}
+
+A common use case for compartments is to have one code base that operates on multiple data sets as found often in web browsers.
+We envision this to be modelled by CID sealing and present our preliminary model for code and data compartments in the following paragraph:
+
+The capabilities for different data sets are separated into compartments.
+There will be one code compartment for each data, but each code capability maps to the same code.
+This means that all code capabilities are identical in all fields, except for the CID field.
+Depending on whether the different compartments trust each other, the compartments can be placed into the same CID space.
+
+It is also possible to use conventional CHERI compartmentalisation where each data compartment has the same CID, but the data capabilities are non-overlapping sets between the compartments.
+In this case, the supervisor code needs to be careful that two data capabilities from different compartments are never accessible to one capability at the same time.
+
+\subsection{Revocation}
+
+When employing more compartments than storable in the ACID space, software will need to virtualise CIDs.
+This will lead to one or more ACIDs needing to be reused.
+In order to maintain safety and security, every capability from the old compartment needs to be made inaccessible.
+This prevents leaking capabilities from the old to the new compartment where both of them have the same ACID, but different SCIDs.
+
+We have come up with a preliminary revocation mechanism, which we will sketch in the following paragraphs.
+Please note that this mechanism likely incurs a substantial performance penalty.
+
+When the supervisor code, e.g., the operating system, has run out of available ACIDs, it needs to revoke a ACID currently in use.
+We currently propose to pick this ACID by random.
+However, one could imagine keeping information in the OS that would enable different strategies, e.g., implementing a least recently used (LRU) policy.
+The OS saves all capabilities of the old compartment in its own space as they are.
+During this sweep, the OS will make all of these capabilities unusable, e.g., by marking it with an \textit{unusable} bit.
+If code tries to legitimately use this capability, the OS will need to jump in and assign a new ACID to this SCID and update all of its pointers to make it usable again.
+For example, a legitimate use case would be waking up the compartment after a longer phase of not invoking it.
+However, we expect there to be many causes for the need of revocation.
```

Comment:
I'm not sure what this sentence means. You mean that revocation already needs to happen for e.g. the heap etc, and it could all be rolled into the same sweep? I'd make that more explicit if it's what you mean.